**contexto**

La empresa quiere predecir el éxito de futuras aplicaciones móviles. Han extraído un volcado de datos de la Google Play Store, pero el pipeline de Machine Learning rechaza el archivo. Su misión como ingenieros de datos es sanear los valores faltantes aplicando reglas de negocio estrictas.

In [18]:
import pandas as pd
import numpy as np
 
 
url = "https://raw.githubusercontent.com/mricardo89/data-mining/main/Unit-II/datasets/googleplaystore.csv"
df = pd.read_csv(url)

In [19]:
df.isnull().sum()


App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

*Ejecuten df.isnull().sum(). Identificarán que la columna Rating tiene casi 1,500 valores nulos, mientras que Type y Android Ver tienen solo 1 o 2.*

### **Pregunta A: ¿Por qué es arquitectónicamente seguro eliminar las filas donde falta Android Ver, pero sería un error fatal eliminar las filas donde falta Rating?**

por el simple hecho que la magnitud es completamente diferente entre borrar solo 3 filas de informacion a borrar casi una 1/10 parte

In [20]:
print(len(df))
df = df.dropna(subset=['Android Ver', 'Type'])
print(len(df))


10841
10837


### **Pregunta B: Verifiquen que el total de registros en el DataFrame apenas disminuyó**

pasó de 10841 a 10837, osea si

In [21]:
medianas = df.groupby('Category')['Rating'].transform('median')
df['Rating'] = df['Rating'].fillna(medianas)

df['Rating'].isnull().sum()

np.int64(0)

*Rellenar los 1,500 nulos de Rating con el promedio global es un error, ya que las apps de la categoría FAMILY (Familia) no se califican igual que las de DATING (Citas).*

### **Pregunta C: Utilizando la lógica de la semana pasada, escriban un bloque combinando .groupby('Category') y .transform() para rellenar los NaN de la columna Rating usando la mediana de su propia categoría.**

In [22]:
df['Size'] = df['Size'].replace('Varies with device', np.nan)

df['Size'].isnull().sum()

np.int64(1694)

*Revisen la columna Size (Tamaño de la app). Notarán que no tiene nulos matemáticos (NaN), pero muchas filas dicen "Varies with device". Para un modelo matemático, ese texto es basura indigerible. Escriban una instrucción usando df['Size'].replace(...) para convertir exactamente esa frase de texto en un nulo matemático real usando np.nan de NumPy.*

### **Pregunta D: ¿Por qué es mejor convertirlo a NaN en esta etapa en lugar de ponerle un tamaño falso como "0 MB"?**

porque 0 MB es un dato falso que el modelo trataría como real, lo que jala la media hacia abajo, osea, sesgo, y no queremos eso.